src/pgen/torus.cpp

```diff
@@ -534,15 +547,115 @@ void PostInitializationModifier(ParameterInput *pin, Mesh *pmesh) {
+    auto gpkg = pmb->packages.Get("geometry");
+    bool derefine_poles = gpkg->Param<bool>("derefine_poles");
+    Real h = gpkg->Param<Real>("h");
+    Real xt = gpkg->Param<Real>("xt");
+    Real alpha = gpkg->Param<Real>("alpha");
+    Real x0 = gpkg->Param<Real>("x0");
+    Real smooth = gpkg->Param<Real>("smooth");
+    auto tr = Geometry::McKinneyGammieRyan(derefine_poles, h, xt, alpha, x0, smooth);
 
     auto ib = pmb->cellbounds.GetBoundsI(IndexDomain::interior);
     auto jb = pmb->cellbounds.GetBoundsJ(IndexDomain::interior);
     auto kb = pmb->cellbounds.GetBoundsK(IndexDomain::interior);
 
     PackIndexMap imap;
-    auto v = rc->PackVariables({fluid_prim::bfield}, imap);
+    std::vector<std::string> vars = {fluid_prim::bfield, fluid_prim::density};
+    auto v = rc->PackVariables(vars, imap);
     const int iblo = imap[fluid_prim::bfield].first;
     const int ibhi = imap[fluid_prim::bfield].second;
+    const int irho = imap[fluid_prim::density].first;
+
+    auto tracer_pkg = pmb->packages.Get("tracers");
+    bool do_tracers = tracer_pkg->Param<bool>("active");
+
+    /* tracer initialization. lots of logic to ensure we only put tracers on disk */
+    if (do_tracers) {
+      auto &sc = pmb->swarm_data.Get();
+      auto &swarm = pmb->swarm_data.Get()->Get("tracers");
+      auto rng_pool = tracer_pkg->Param<RNGPool>("rng_pool");
+      const auto num_tracers_total = tracer_pkg->Param<int>("num_tracers");
+      auto &x = swarm->Get<Real>("x").Get();
+      auto &y = swarm->Get<Real>("y").Get();
+      auto &z = swarm->Get<Real>("z").Get();
+      auto &mass = swarm->Get<Real>("mass").Get();
+      auto &id = swarm->Get<int>("id").Get();
+
+      /* lambda to flatten (k, j, i) -> n */
+      auto flatten = KOKKOS_LAMBDA(const int k, const int j, const int i) {
+        return (i + (ib.e - ib.s + 1) * (j + (jb.e - jb.s + 1) * k));
+      };
+
+      // TODO (BLB): better way to construct & access indices array?
+      // Access as indices( flatten( k-kb.s, j-jb.s, i-ib.s ) )
+      const int mysize = (kb.e - kb.s + 1) * (jb.e - jb.s + 1) * (ib.e - ib.s + 1);
+      ParArray1D<int> indices("map from k,j,i to 1D particle array", mysize);
+      int num_cells_disk = 0;
+      /* sum cells on disk and create indices array for mapping (k, j, i) -> n */
+      pmb->par_reduce(
+          "Phoebus::ProblemGenerator::Torus::SumDiskCells", kb.s, kb.e, jb.s, jb.e, ib.s,
+          ib.e,
+          KOKKOS_LAMBDA(const int k, const int j, const int i, int &n_cells_disk) {
+            const Real x1 = coords.Xc<1>(k, j, i);
+            const Real x2 = coords.Xc<2>(k, j, i);
+            Real r = tr.bl_radius(x1);
+            Real th = tr.bl_theta(x1, x2);
+
+            Real lnh = -1.0;
+            Real hm1;
+            Real uphi;
+            if (r > rin) lnh = log_enthalpy(r, th, a, rin, angular_mom, uphi);
+            if (lnh > 0.0 && x1 > xh) {
+              indices(flatten(k - kb.s, j - jb.s, i - ib.s)) = n_cells_disk;
```

Comment:
`n_cells_disk` is not guaranteed to be assigned  in any particular order... so making `indices` have atomic writes may not solve this problem. But let's try that first.